Fix metrics offline engine

dalf opened the pull request:
## What does this PR do?

Fix offline engine:
* don't crash on time recording
* fix `/preferences` and `/stats`

## Why is this change important?

Bug fixes

## How to test this PR locally?

* enable `git grep` engine :
   * send one or few queries
   * check the `/preferences` page: 
      * the total time is displayed (0.0 most probably)
      * no bar chart
   * check the `/stats` page : 
      * processing time  = total time
      * HTTP time column is empty
      * no bar chart
* hack `searx/engines/command.py`: add `import time; time.sleep(1.5)` in `def search(query, params)`
   * send one or few queries
   * check the `/preferences` page (same tests as above, except this time there is a bar chart)
   * check the `/stats` page (same tests as above, except this time there is a bar chart)

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

Related to https://github.com/searxng/searxng/pull/80#issuecomment-843362260

dalf committed:
[fix] offline engine: don't crash on time recording

dalf committed:
[fix] metrics: processing time = total time if there is no http time

It was previsouly None

Fix /stats

dalf committed:
[fix] offline engines: fix templates /stats and /preferences

return42 commented on searx/webapp.py:
```diff
@@ -923,9 +923,9 @@ def preferences():
-            'rate95': rate95 if rate95 else None,
+            'time': median if median is not None else None,
+            'rate80': rate80 if rate80 is not None else None,
+            'rate95': rate95 if rate95 is not None else None,
```
If None, Replace None by None / see 52b5040

return42 commented on searx/network/__init__.py:
```diff
@@ -53,14 +53,14 @@ def set_timeout_for_thread(timeout, start_time=None):
 
 def set_context_network_name(network_name):
-    THREADLOCAL.network = get_network(network_name)
+    THREADLOCAL.__dict__['network'] = get_network(network_name)
```
Using `__dict__`  is only needed in _getters_ when you want None if unset, but should not used in _setters_ / see 08f86c3

return42 commented on searx/metrics/__init__.py:
```diff
@@ -157,7 +157,11 @@ def get_reliabilities(engline_name_list, checker_results):
+    return round(number, digits) if number is not None else number
 
 
 def get_engines_stats(engine_name_list):
```
This function was a little bit hard for me to read, I send a patch trying to make it more readable / see 
09d40dd

return42 committed:
[refactor] metrics.get_reliabilities() - make code more readable

- init stat values by None
- drop round_or_none
- don't try to get percentage if base is 'None'

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/results.py:
```diff
@@ -371,11 +371,13 @@ def add_unresponsive_engine(self, engine_name, error_type, error_message=None, s
-            'load': page_load_time
-        })
+        }
+        if page_load_time or True:
```
Condition is always True / see d9f4eba

return42 reviewed (approved):
I tested this PR on top of #80 (https://github.com/return42/searxng/commits/fix-metrics-offline-engine) and it works ("demo" engine in the screenshot below) .

I also placed some patches on top of your fix-metrics-offline-engine .. theses patches are suggestions, if you have a doubt, drop them. 

![image](https://user-images.githubusercontent.com/554536/119124008-b64a3080-ba1f-11eb-97a8-a57229c5e5c6.png)
